### PR TITLE
Save local run's trace to cosmosDB when workspace info is configured

### PR DESCRIPTION
**Pull request by huaiyan**
# Description

Add logic to write trace to cosmosDB when customer defined workspace info by `pf config set trace.provider=azureml:/subscriptions/<subscription_id>/resourceGroups/<resource_group>/providers/Microsoft.MachineLearningServices/workspaces/<workspace_name>`

How do this work:
We register `trace_collector` url as `/v1/traces` in `promptflow/_sdk/_service/app`.
As handling `trace_collector`, if there is workspace info set, we try to create/reuse cosmosDB client and write trace data to cosmosDB.

For now, each line's execution will submit `/v1/traces` request, aggregate node's execution also will submit one separate `/v1/traces` request.

Known issue: 
1. The operation to init cosmosDB client could be time consuming (Usually about 20 seconds), we cache the token for 4 minutes, so sometimes it will be pretty fast.
2.  Due to the time-consuming issue above, the order to insert main item and eval dict to `LineSummary` may be wrong. Add 1 second sleeping for inserting eval result.

TODOs @zhengfeiwang: SDK may need to further improve the client's acquirement (use default credential for example)

# All Promptflow Contribution checklist:
- [X] **The pull request does not introduce [breaking changes].**
- [ ] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [X] **I have read the [contribution guidelines](../CONTRIBUTING.md).**
- [ ] **Create an issue and link to the pull request to get dedicated review from promptflow team. Learn more: [suggested workflow](../CONTRIBUTING.md#suggested-workflow).**

## General Guidelines and Best Practices
- [X] Title of the pull request is clear and informative.
- [X] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### Testing Guidelines
- [X] Pull request includes test coverage for the included changes.

### CODE REVIEW

File: src/promptflow/promptflow/_sdk/_service/apis/collector.py [hunk 1]
```diff
@@ -37,9 +60,23 @@ def trace_collector():
...
+                        _write_trace_to_cosmosdb(span, line_summary_client, span_client)
         return "Traces received", 200
 
     # JSON protobuf encoding
     elif "application/json" in content_type:
         raise NotImplementedError
+
+
+def _write_trace_to_cosmosdb(span: Span, line_summary_client, span_client):
+    from promptflow.azure._storage.cosmosdb.span import Span as SpanCosmosDB
```
Comment (thy09) [inline src/promptflow/promptflow/_sdk/_service/apis/collector.py:75]: Can these work when pf[azure] is not installed but the environment variables are set?

File: src/promptflow/promptflow/azure/_storage/cosmosdb/summary.py [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+        else:
+            cumulative_token_count = None
+        item = SummaryLine(
+            id=self.span.trace_id,  # trace id is unique for LineSummary container
+            partition_key=session_id,
+            session_id=session_id,
+            trace_id=self.span.trace_id,
+            root_span_id=self.span.span_id,
+            inputs=json.loads(attributes[SpanAttributeFieldName.INPUTS]),
+            outputs=json.loads(attributes[SpanAttributeFieldName.OUTPUT]),
```
Comment (zhengfeiwang) [inline src/promptflow/promptflow/azure/_storage/cosmosdb/summary.py:119]: what will happen if there is `float("nan")` in inputs/outputs? as they are not standard JSON value, and we have observed issues with those in local, and the solution is to hanlde in SDK side.

File: src/promptflow/promptflow/_sdk/_service/apis/collector.py [hunk 3]
```diff
@@ -18,8 +19,30 @@
...
+    )
+
+
 def trace_collector():
     content_type = request.headers.get("Content-Type")
+    if _is_workspace_configured():
+        from promptflow.azure._storage.cosmosdb.client import get_client
+
+        subscription_id, resource_group_name, workspace_name = _get_workspace_info()
+        span_client = get_client("Span", subscription_id, resource_group_name, workspace_name)
```
Comment (zhengfeiwang) [inline src/promptflow/promptflow/_sdk/_service/apis/collector.py:44]: try to make these constants defined in common constant file?

File: src/promptflow/promptflow/azure/_storage/cosmosdb/span.py [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+        self.links = span._content[SpanFieldName.LINKS]
+        self.resource = span._content[SpanFieldName.RESOURCE]
+        self.partition_key = span.session_id
+        self.id = span.span_id
+
+    def persist(self, client):
+        if self.id is None or self.partition_key is None or self.resource is None:
+            return
+
+        resource_attributes = self.resource.get("attributes", None)
```
Comment (zhengfeiwang) [inline src/promptflow/promptflow/azure/_storage/cosmosdb/span.py:47]: we already defined some consts for attributes name, try to replace this hard code to that one

File: src/promptflow/promptflow/azure/_storage/cosmosdb/summary.py (hunk 2 above)
Comment (huaiyan) [reply src/promptflow/promptflow/azure/_storage/cosmosdb/summary.py:119]: Thanks a lot! For current code, cosmosDB will raise Exception for badrequest.

File: src/promptflow/promptflow/azure/_storage/cosmosdb/span.py (hunk 4 above)
Comment (riddlexu) [reply src/promptflow/promptflow/azure/_storage/cosmosdb/span.py]: Fixed. Thanks.

File: src/promptflow/promptflow/_sdk/_service/apis/collector.py (hunk 3 above)
Comment (huaiyan) [reply src/promptflow/promptflow/_sdk/_service/apis/collector.py]: Fixed~

**Review by riddlexu — DISMISSED** [review]

**Review by thy09 — DISMISSED** [review]

File: src/promptflow/promptflow/_sdk/_service/apis/collector.py [hunk 5]
```diff
@@ -64,19 +44,32 @@ def trace_collector():
...
 
-def _write_trace_to_cosmosdb(span: Span, line_summary_client, span_client):
+def _try_write_trace_to_cosmosdb(span: Span):
+    span_resource = span._content.get(SpanFieldName.RESOURCE, {})
+    resource_attributes = span_resource.get(SpanResourceFieldName.ATTRIBUTES, {})
+    subscription_id = resource_attributes.get(SpanResourceAttributesFieldName.SUBSCRIPTION_ID, None)
+    resource_group_name = resource_attributes.get(SpanResourceAttributesFieldName.RESOURCE_GROUP_NAME, None)
+    workspace_name = resource_attributes.get(SpanResourceAttributesFieldName.WORKSPACE_NAME, None)
+    if subscription_id is None or resource_group_name is None or workspace_name is None:
+        current_app.logger.info("Cannot find workspace info in span resource, skip writing to cosmosdb.")
```
Comment (zhengfeiwang) [inline src/promptflow/promptflow/_sdk/_service/apis/collector.py:62]: maybe debug is better? otherwise, there will be many info lines for this?

Comment (huaiyan) [reply src/promptflow/promptflow/_sdk/_service/apis/collector.py]: Fixed

**Review by zhengfeiwang — APPROVED** [review]

**Review by thy09 — APPROVED** [review]

**Pull request merged**